core/foundation/inc/ROOT/TypeTraits.hxx

```diff
@@ -0,0 +1,203 @@
+// @(#)root/foundation:
+// Author: Axel Naumann, Enrico Guiraud, June 2017
+
+/*************************************************************************
+ * Copyright (C) 1995-2017, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_TypeTraits
+#define ROOT_TypeTraits
+
+#include <memory> // shared_ptr, unique_ptr for IsSmartOrDumbPtr
+#include <type_traits>
+#include <vector> // for IsContainer
+
+// fwd declaration for IsV7Hist
+class TH1;
+
+namespace ROOT {
+
+// fwd declaration for IsV7Hist
+namespace Experimental {
+template <int D, typename P, template <int, typename, template <typename> class> class... S>
+class THist;
+} // ns Experimental
+
+/// ROOT type_traits extensions
+namespace TypeTraits {
+
+///\class ROOT::TypeTraits::
+template <class T>
+class IsSmartOrDumbPtr : public std::integral_constant<bool, std::is_pointer<T>::value> {
+};
+
+template <class P>
+class IsSmartOrDumbPtr<std::shared_ptr<P>> : public std::true_type {
+};
+
+template <class P>
+class IsSmartOrDumbPtr<std::unique_ptr<P>> : public std::true_type {
+};
+
+/// Check for container traits.
+template <typename T>
+struct IsContainer {
+   using Test_t = typename std::decay<T>::type;
+
+   template <typename A>
+   static constexpr bool Test(A *pt, A const *cpt = nullptr, decltype(pt->begin()) * = nullptr,
```

Comment:
Thanks for looking into this. This was implemented by @dpiparo so I would ask his opinion too.